File: cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<vector<pair<int, int>>> adj(n);
        for(auto flight : flights){
            adj[flight[0]].emplace_back(flight[1], flight[2]);
        }
        queue<pair<int, pair<int, int>>> q;
        vector<int> dist(n, INT_MAX);

        q.push({src, {0, 0}});
        dist[src] = 0;
        while(!q.empty()){
            auto cur = q.front();
            q.pop();
            int node = cur.first;
            int price = cur.second.first;
            int stops = cur.second.second;

            if(stops > k) continue;

            for(auto i : adj[node]){
                int next = i.first;
                int cost = i.second;

                if(price + cost < dist[next]){
                    dist[next] = price + cost;
                    q.push({next, {dist[next], stops + 1}});
                }
            }
        }
        return dist[dst] == INT_MAX ? -1 : dist[dst];
    }
};

File: cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: list[list[int]], src: int, dst: int, k: int) -> int:
        adj = [[] for _ in range(n)]
        for fr, to, cost in flights:
            adj[fr].append((to, cost)) # adj list

        dist = [[float("inf")] * (k + 10) for _ in range(n)] # dist storing k+2 cols and n rows
                                                # dist[n][c] gives cheapest way to get to node n in c flights
        dist[src][0] = 0 # getting to src node in 0 flights is free
        minheap = [(0, src, 0)] # minheap pqueue, cost, node, flights taken

        while minheap:
            cost, node, flights = heapq.heappop(minheap)
            if node == dst: return cost # found cheapest
            if flights == k+1 or dist[node][flights] < cost: continue # no more stops or stopping here is worse

            for nei, w in adj[node]:
                newcost = cost + w
                newflights = flights+1
                # check if neighbours are cool
                if dist[nei][newflights] > newcost:
                    dist[nei][newflights] = newcost
                    heapq.heappush(minheap, (newcost, nei, newflights)) # only push to mh if neighbour actually improves, since pushing a worse state (more cost) will never be better

        return -1

